File: problems/CSES/Graph_Algorithms/de_bruijn_sequence/de_bruijn_sequence.py
import sys
sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n = int(input())

    if n == 1:
        print("01")
        return

    size = 1 << (n - 1)
    mask = size - 1

    used = [0] * size
    res = []

    def dfs(v):
        while used[v] < 2:
            bit = used[v]
            used[v] += 1

            u = ((v << 1) | bit) & mask

            dfs(u)
            res.append(str(bit))

    dfs(0)

    res.reverse()
    print("0" * (n - 1) + "".join(res))

main()

File: problems/CSES/Graph_Algorithms/de_bruijn_sequence/de_bruijn_sequence.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

void solve() {
    return;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    if (n == 1) {
        cout << "01\n";
        return 0;
    }

    int size = 1 << (n - 1);
    int mask = size - 1;

    vector<int> used(size, 0);
    string res;

    function<void(int)> dfs = [&](int v) {
        while (used[v] < 2) {
            int bit = used[v];
            ++used[v];

            int u = ((v << 1) | bit) & mask;

            dfs(u);
            res.push_back('0' + bit);
        }
    };

    dfs(0);

    reverse(res.begin(), res.end());

    cout << string(n - 1, '0') << res << '\n';

    return 0;
}
